src/routes/(blog-article)/containerizing-rust-applications-best-practices/+page.md

```diff
@@ -0,0 +1,1049 @@
+WORKDIR /build/src
+COPY --from=recipe /build/recipe.json /build/recipe.json
+RUN cargo chef cook --tests --benches --examples --workspace --all-targets --all-features --recipe-path /build/recipe.json --release
+RUN cargo nextest archive --tests --benches --examples --workspace --all-targets --all-features --archive-file /build/temp.tar.zst --release  ; rm -f /build/temp.tar.zst
+```
+
+</CodeBlock>
+
+In the following stage we build the application. The `cargo nextest archive` command builds
+and archives the application.
+
+<CodeBlock lang="dockerfile">
+
+```dockerfile
+## Build Archive (release)
+FROM dependencies AS build
+WORKDIR /build/src
+COPY . /build/src
+RUN cargo nextest archive --tests --benches --examples --workspace --all-targets --all-features --archive-file /build/full-example.tar.zst --release
+```
+
+</CodeBlock>
+
+Now, that we have successfully built the application, we can run the tests. We
+extract the application from the archived artifacts and run the tests.
+
+<CodeBlock lang="dockerfile">
+
+```dockerfile
+## Extract and Test (release)
+FROM tester as test
+WORKDIR /test
+COPY . /test/src
+COPY --from=build \
+  /build/full-example.tar.zst \
+  /test/full-example.tar.zst
+# It compiles without running the tests to make sure there are no problems with
+# the archive.
+RUN cargo nextest run --workspace-remap /test/src/ --extract-to /test/src/ --no-run --archive-file /test/full-example.tar.zst
+# We actually run the tests
+RUN cargo nextest run --workspace-remap /test/src/ --target-dir-remap /test/src/target/ --cargo-metadata /test/src/target/nextest/cargo-metadata.json --binaries-metadata /test/src/target/nextest/binaries-metadata.json
+# We copy the application binary to an specified location so we can get it from
+# there in the final runtime stage.
+RUN mkdir -p /app/bin/; cp -l /test/src/target/release/full-example /app/bin/full-example
+# Since we use su-exec. We need to run the container as root.
+RUN chown -R root:root /app; chmod -R u=rw,go=r,a+X /app; chmod -R a+x /app/bin
+```
+
+</CodeBlock>
+
+Once the application has been built and tested we prepare the runtime. We start
+from a minimum "distroless" image variant. We add an entrypoint to setup the application
+and also to make sure we don't use the `root` user to run it. The entrypoint just
+run the application provided as an argument, in our case, our application in `debug` or
+`release` mode, depending of which one you want to run.
+
+<CodeBlock lang="dockerfile">
+
+```dockerfile
+## Runtime
+FROM gcr.io/"distroless/cc-debian12:debug as runtime
+RUN ["/busybox/cp", "-sp", "/busybox/sh","/busybox/cat","/busybox/ls","/busybox/env", "/bin/"]
+COPY --from=gcc --chmod=0555 /usr/local/bin/su-exec /bin/su-exec
+ARG USER_ID=1000
+COPY --chmod=0555 ./share/container/entry_script_sh /usr/local/bin/entry.sh
+ENTRYPOINT ["/usr/local/bin/entry.sh"]
+```
+
+</CodeBlock>
+
+Finally, we add the `CMD` instruction so we can execute the application. The
```

Comment:
Why are these lines and some of the following which seem to be text in blue here?

Reply:
They are a `key` words for Dockerfile instructions or values.